src/sentry/api/endpoints/organization_stats_v2.py

```diff
@@ -0,0 +1,239 @@
+        self.other = StatMeasure(0, 0)
+        self.time = time
+
+    def serialize(self):
+        return {
+            "accepted": {
+                "quantity": self.accepted.quantity,
+                "times_seen": self.accepted.times_seen,
+            },
+            "filtered": {"quantity": self.filtered.quantity, "times_seen": self.filtered.quantity},
+            "dropped": {
+                "overQuota": {
+                    "quantity": self.over_quota.quantity,
+                    "times_seen": self.over_quota.quantity,
+                },
+                "spikeProtection": {
+                    "quantity": self.spike_protection.quantity,
+                    "times_seen": self.spike_protection.quantity,
+                },
+                "other": {"quantity": self.other.quantity, "times_seen": self.other.quantity},
+            },
+            "time": self.time,
+        }
+
+    def update(self, row):
+        if row["outcome"] == Outcome.RATE_LIMITED:
+            if row["reason"] in {"usage_exceeded", "grace_period"}:
+                self.over_quota.quantity = row["quantity"]
+                self.over_quota.times_seen = row["times_seen"]
+            elif row["reason"] in "smart_rate_limit":
+                self.spike_protection.quantity = row["quantity"]
+                self.spike_protection.times_seen = row["quantity"]
+            else:
+                self.other.quantity = row["quantity"]
+                self.other.times_seen = row["quantity"]
+
+        elif row["outcome"] == Outcome.ACCEPTED:
+            self.accepted.quantity = row["quantity"]
+            self.accepted.times_seen = row["times_seen"]
+        elif row["outcome"] == Outcome.FILTERED:
+            self.filtered.quantity = row["quantity"]
+            self.filtered.times_seen = row["times_seen"]
+
+    def rate_limited_reason_mapper(reason):
+        # billing logic ported over. TODO: combine usages of this into some abstracted module?
+        if reason in {"usage_exceeded", "grace_period"}:
+            reason_val = "overQuota"
+        elif reason == "smart_rate_limit":
+            reason_val = "spikeProtection"
+        else:
+            reason_val = "other"
+        return reason_val
+
+
+# TODO: verify what kind of timestamp we return to frontend
+def zerofill(data, start, end, rollup, orderby):
+    rv = {}
+    start = int(to_naive_timestamp(naiveify_datetime(start)) / rollup) * rollup
+    end = (int(to_naive_timestamp(naiveify_datetime(end)) / rollup) * rollup) + rollup
+    data_by_time = {}
+
+    for obj in data:
+        if obj["time"] in data_by_time:
+            data_by_time[obj["time"]].append(obj)
+        else:
+            data_by_time[obj["time"]] = [obj]
+    for key in range(start, end, rollup):
+        if key in data_by_time and len(data_by_time[key]) > 0:
+            rv = rv + data_by_time[key]
+            data_by_time[key] = []
+        else:
+            val = MeasureValue(key)
+            rv[key] = val
```

Comment:
It seems like `zerofill` is doing a lot of work up front and it's a little hard to follow. Is it only filling in empty entries in case the corresponding timestamps don't later show up in the data? If so, is there a good way to push that to the serialization step, starting instead with empty lists/dicts?

Just a thought. I'll take your word for it if you say that's more trouble than it's worth.

Reply:
yeah, how it is now isn't ideal. i think ideally we move it into the outcomes snuba query, and this code can gracefully create an object and this complexity can be removed from the class.